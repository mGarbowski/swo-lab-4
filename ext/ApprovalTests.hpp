#ifndef APPROVAL_TEST_1_APPROVALTESTS_HPP
#define APPROVAL_TEST_1_APPROVALTESTS_HPP
#include "ApprovalTests.v.10.13.0.hpp"
#endif
