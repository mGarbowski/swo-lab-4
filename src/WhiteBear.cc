#include "WhiteBear.h"
#include <sstream>
#include <memory>


void WhiteBear::updateQuality() {
    for (auto &item: items_) {
        item->updateQuality();
    }
}

void WhiteBear::addItem(UItem item) { items_.push_back(std::move(item)); }

void WhiteBear::printItems() {
    for (auto &item: items_) {
        std::cout << *item << std::endl;
    }
}

void WhiteBear::printItems(std::ostream &output) {
    for (auto &item: items_) {
        output << *item << std::endl;
    }
    output << std::endl;
}

