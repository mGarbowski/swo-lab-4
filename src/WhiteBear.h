#ifndef WHITE_BEAR_H
#define WHITE_BEAR_H

#include <iostream>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "items/Item.h"

class WhiteBear {
 public:
  void printItems();

  void printItems(std::ostream &output);

  void addItem(UItem item);

  void updateQuality();

 private:
  std::vector<UItem> items_;
};

#endif  // WHITE_BEAR_H
