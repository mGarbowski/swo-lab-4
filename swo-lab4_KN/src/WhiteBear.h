#ifndef WHITE_BEAR_H
#define WHITE_BEAR_H
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class WhiteBear;

class Item {
public:
  Item(string name, int days_remaining, int quality)
      : t(name), d(days_remaining), v(quality) {}
  friend class WhiteBear;
  friend std::ostream &operator<<(std::ostream &s, Item &item);

private:
  string t;
  int d;
  int v;
};

class WhiteBear {
public:
  void printItems();
  void printItems(std::ostream& output);
  void addItem(const Item &item);
  void updateQuality();

private:
  vector<Item> items_;
};

const std::string item1 = "Cheese Brie";
const std::string item2 = "Tickets to a concert";
const std::string item3 = "Legolas, Hand of Gollum";
#endif // WHITE_BEAR_H
