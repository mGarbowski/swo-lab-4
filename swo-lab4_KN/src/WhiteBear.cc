#include "WhiteBear.h"
#include <sstream>

std::string Item::toString() const {
    stringstream result;
    result << name << ", " << daysRemaining << ", " << quality;
    return result.str();
}

std::ostream &operator<<(std::ostream &s, Item &item) {
  s << item.toString();
  return s;
}

/**
const std::string CHEESE_BRIE = "Cheese Brie";
const std::string TICKETS = "Tickets to a concert";
const std::string LEGOLAS = "Legolas, Hand of Gollum";
*/

void WhiteBear::updateQuality() {
  for (auto & item : items_) {
    if (item.name != CHEESE_BRIE && item.name != TICKETS) {
      if (item.quality > 0) {
        if (item.name != LEGOLAS) {
          item.quality -= 1;
        }
      }
    } else {
      if (item.quality < 50) {
        item.quality = item.quality + 1;
        if (item.name == TICKETS) {
          if (item.daysRemaining < 11) {
            if (item.quality < 50) {
              item.quality += 1;
            }
          }
          if (item.daysRemaining < 6) {
            if (item.quality < 50) {
              item.quality++;
            }
          }
        }
      }
    }
    if (item.name != LEGOLAS) {
      --item.daysRemaining;
    }
    if (item.daysRemaining < 0) {
      if (item.name != CHEESE_BRIE) {
        if (item.name != TICKETS) {
          if (item.quality > 0) {
            if (item.name != LEGOLAS) {
              item.quality = item.quality - 1;
            }
          }
        } else {
          item.quality = item.quality - item.quality;
        }
      } else {
        if (item.quality < 50) {
          ++item.quality;
        }
      }
    }
  }
}

void WhiteBear::addItem(const Item &item) { items_.push_back(item); }

void WhiteBear::printItems() {
  for (auto & item : items_) {
    std::cout << item << std::endl;
  }
}

void WhiteBear::printItems(std::ostream& output) {
  for (auto & item : items_) {
    //output << *i << std::endl;
	std::string s;
	output << item.name << ", " << item.daysRemaining << ", " << item.quality << std::endl;
  }
  output << std::endl;
}

