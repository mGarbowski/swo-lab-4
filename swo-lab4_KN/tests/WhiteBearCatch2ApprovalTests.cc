#include "catch2/catch_all.hpp"
#include "ApprovalTests.hpp"
#include "WhiteBear.h"


TEST_CASE("WhiteBearApprovalTests", "First Test") {

    WhiteBear app;

    app.addItem(Item("+7 Yellow Vest", 10, 20));
    app.addItem(Item("Cheese Brie", 10, 20));
    app.addItem(Item("Tickets to a concert", 10, 20));
    app.addItem(Item("Legolas, Hand of Gollum", 10, 20));

    std::stringstream output;

    for (int day = 0; day <= 20; ++day) {
        output << "Day is " << day << std::endl;
        app.updateQuality();
        app.printItems(output);
    }

    ApprovalTests::Approvals::verify(output.str());
}

