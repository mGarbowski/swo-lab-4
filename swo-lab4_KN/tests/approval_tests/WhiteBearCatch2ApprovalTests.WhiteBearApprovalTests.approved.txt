Day is 0
+7 Yellow Vest, 9, 19
Cheese Brie, 9, 21
Tickets to a concert, 9, 22
Legolas, Hand of Gollum, 10, 20

Day is 1
+7 Yellow Vest, 8, 18
Cheese Brie, 8, 22
Tickets to a concert, 8, 24
Legolas, Hand of Gollum, 10, 20

Day is 2
+7 Yellow Vest, 7, 17
Cheese Brie, 7, 23
Tickets to a concert, 7, 26
Legolas, Hand of Gollum, 10, 20

Day is 3
+7 Yellow Vest, 6, 16
Cheese Brie, 6, 24
Tickets to a concert, 6, 28
Legolas, Hand of Gollum, 10, 20

Day is 4
+7 Yellow Vest, 5, 15
Cheese Brie, 5, 25
Tickets to a concert, 5, 30
Legolas, Hand of Gollum, 10, 20

Day is 5
+7 Yellow Vest, 4, 14
Cheese Brie, 4, 26
Tickets to a concert, 4, 33
Legolas, Hand of Gollum, 10, 20

Day is 6
+7 Yellow Vest, 3, 13
Cheese Brie, 3, 27
Tickets to a concert, 3, 36
Legolas, Hand of Gollum, 10, 20

Day is 7
+7 Yellow Vest, 2, 12
Cheese Brie, 2, 28
Tickets to a concert, 2, 39
Legolas, Hand of Gollum, 10, 20

Day is 8
+7 Yellow Vest, 1, 11
Cheese Brie, 1, 29
Tickets to a concert, 1, 42
Legolas, Hand of Gollum, 10, 20

Day is 9
+7 Yellow Vest, 0, 10
Cheese Brie, 0, 30
Tickets to a concert, 0, 45
Legolas, Hand of Gollum, 10, 20

Day is 10
+7 Yellow Vest, -1, 8
Cheese Brie, -1, 32
Tickets to a concert, -1, 0
Legolas, Hand of Gollum, 10, 20

Day is 11
+7 Yellow Vest, -2, 6
Cheese Brie, -2, 34
Tickets to a concert, -2, 0
Legolas, Hand of Gollum, 10, 20

Day is 12
+7 Yellow Vest, -3, 4
Cheese Brie, -3, 36
Tickets to a concert, -3, 0
Legolas, Hand of Gollum, 10, 20

Day is 13
+7 Yellow Vest, -4, 2
Cheese Brie, -4, 38
Tickets to a concert, -4, 0
Legolas, Hand of Gollum, 10, 20

Day is 14
+7 Yellow Vest, -5, 0
Cheese Brie, -5, 40
Tickets to a concert, -5, 0
Legolas, Hand of Gollum, 10, 20

Day is 15
+7 Yellow Vest, -6, 0
Cheese Brie, -6, 42
Tickets to a concert, -6, 0
Legolas, Hand of Gollum, 10, 20

Day is 16
+7 Yellow Vest, -7, 0
Cheese Brie, -7, 44
Tickets to a concert, -7, 0
Legolas, Hand of Gollum, 10, 20

Day is 17
+7 Yellow Vest, -8, 0
Cheese Brie, -8, 46
Tickets to a concert, -8, 0
Legolas, Hand of Gollum, 10, 20

Day is 18
+7 Yellow Vest, -9, 0
Cheese Brie, -9, 48
Tickets to a concert, -9, 0
Legolas, Hand of Gollum, 10, 20

Day is 19
+7 Yellow Vest, -10, 0
Cheese Brie, -10, 50
Tickets to a concert, -10, 0
Legolas, Hand of Gollum, 10, 20

Day is 20
+7 Yellow Vest, -11, 0
Cheese Brie, -11, 50
Tickets to a concert, -11, 0
Legolas, Hand of Gollum, 10, 20


